### PR TITLE
feat(users): enforce case-insensitive unique email on registration

### LINKED ISSUE #1625: [TASK] API - enforce one account per email on registration
## Problem / Context

Registration must guarantee one account per email. `apps/api/src/auth/registration.service.ts` creates users on self-service signup; we need to confirm and enforce that a duplicate email cannot create a second account, and that the failure returns a clear, non-enumerating-but-actionable error to the FE `register-form`.

## Proposed Solution

- Enforce email uniqueness at registration (DB-level unique constraint on the users table plus a service-level check that maps the violation to a clean domain error), following the repository error-handling pattern (catch infrastructure unique-violation -> throw a domain exception).
- Return a clear FE-facing error on the `POST /auth/register` path so `apps/web/src/features/users/components/register-form.tsx` can surface "an account with this email already exists".
- Verify behavior interacts correctly with the email-confirmation flow (e.g. an unconfirmed duplicate signup should not silently create a second row).

## Classification

**Type**: API (+ FE error copy)
**Layer**: Application / Infrastructure (auth, users)
**File(s)**: `apps/api/src/auth/registration.service.ts`, users ORM entity + migration (`apps/api/src/**` / `libs/core/src/users/**`), `apps/web/src/features/users/components/register-form.tsx`

## Dependencies

- Coordinates with the email-confirmation sibling issue (duplicate/unconfirmed handling)

## Assumptions

- Email is the account identity key; a unique constraint is acceptable (add a migration if one does not already exist).

## Acceptance Criteria

- [ ] A second registration with an existing email is rejected (DB unique constraint + domain error)
- [ ] The FE register form shows a clear "email already in use" message
- [ ] No duplicate rows are created under the confirmation flow (unconfirmed re-signup handled sanely)
- [ ] Migration added if the unique constraint is new; `migration:show` clean
- [ ] Tests added for the duplicate-email path
- [ ] No architecture boundary violations (CORE <-> Integration)

### PR DESCRIPTION
**Pull request by jakubretajczykBD**
Normalizes email (trim + lowercase) in UserRepository.findByEmail/save so foo@example.com and Foo@Example.com collide against the same UQ_users_email constraint, backfills existing rows via a new migration, and surfaces a dedicated "This email is already registered." message on the register form for 409 responses.

## Summary
- Normalizes `users.email` to trimmed-lowercase in `UserRepository.findByEmail`/`save` so `foo@example.com` and `Foo@Example.com` collide against the existing `UQ_users_email` constraint instead of creating two accounts.
- Adds migration `1819000000000-normalize-user-emails.ts` to backfill any pre-existing mixed-case rows (idempotent; fails loudly on a genuine pre-existing collision rather than silently merging accounts).
- Adds `ApiError.isConflict()` and surfaces a dedicated "This email is already registered." message on the register form for 409 responses, instead of the raw backend error string.
- The DB unique constraint, service-level pre-check, and repository-level 409 mapping already existed prior to this change — this PR closes the one real gap (case sensitivity).

## Related issues

Part of #1606
Closes #1625

## Test plan
- [x] `pnpm --filter @openlinker/core test -- user.repository.spec.ts` — normalization + concurrent-registration-race path (Postgres `23505` → `UserAlreadyExistsException`)
- [x] `pnpm --filter @openlinker/api test -- registration.service.spec.ts password-reset.service.spec.ts` — existing duplicate-email/username paths unaffected
- [ ] `pnpm --filter @openlinker/web test -- register-form api-error` — FE conflict-message + `isConflict()` unit tests
- [ ] `pnpm --filter @openlinker/api migration:run` against a fresh DB
- [ ] Manual: register `test@example.com`, then `Test@Example.com` → expect "This email is already registered."

## Quality gate

- [ ] `pnpm lint` passes (zero errors)
- [ ] `pnpm type-check` passes (zero errors)
- [ ] `pnpm test` passes (all unit tests green)
- [ ] *Optional, Docker required:* `pnpm test:integration` passes — needed
      only if you touched `apps/api/test/integration/**` or any plugin's
      `infrastructure/adapters/`.

## Migrations

- [ ] If this PR changes an ORM entity, a migration is included under
      `apps/api/src/migrations/` (or the plugin package), `pnpm --filter
      @openlinker/api migration:show` confirms it's listed, and both
      `up()` and `down()` were tested locally. See
      [`docs/migrations.md`](../docs/migrations.md). Tick this box for
      PRs that don't touch schemas too — it's trivially satisfied.
- Adds migration `1819000000000-normalize-user-emails.ts` to backfill any pre-existing mixed-case rows (idempotent; fails loudly on a genuine pre-existing collision rather than silently merging accounts).
- 

---

<sub>By submitting this pull request, I confirm that my contributions
are made under the terms of the [Apache License 2.0](../LICENSE), and I
certify the [Developer Certificate of Origin](https://developercertificate.org/)
by signing off my commits.</sub>

### CODE REVIEW

**Comment by norbert-kulus-blockydevs** [comment]
## Code review: #1641 (case-insensitive unique email on registration)

Summary: this PR normalizes `email` to trimmed-lowercase at every read/write path in `UserRepository` (`findByEmail`, `save`), backfills existing rows via a new migration, and adds a dedicated 409 message on the FE register form. The core fix is small, correctly targeted, and genuinely closes the TOCTOU gap - not just an app-level check.

### Verification of the 6 specific risk areas from the issue

1. **DB-level enforcement, not TOCTOU-vulnerable**: confirmed safe. `UserRepository.save` normalizes the email *before* building the insert entity, and the existing `UQ_users_email` unique constraint (`apps/api/src/migrations/1775000000000-add-users-table.ts`) enforces exact-string uniqueness on that now-always-lowercased value. Two concurrent registrations with `foo@example.com` / `Foo@Example.com` will both pass `RegistrationService`'s pre-check (neither exists yet), but the second `INSERT` will hit the real Postgres unique-violation on the normalized string, get caught in `save()`'s existing `23505` handler, and surface as `UserAlreadyExistsException` -> `ConflictException` (409). This is the correct normalize-on-write pattern (not `citext`, but equivalent here since all writes route through this one repository).
2. **Migration handles pre-existing case-duplicates correctly**: `1819000000000-normalize-user-emails.ts` runs a targeted `UPDATE ... WHERE email <> LOWER(TRIM(email))`. If two existing rows would normalize to the same value, the `UPDATE` fails loudly with a `23505` against the existing `UQ_users_email` constraint rather than silently merging accounts - documented in the migration's header and in the PR description. This is the right failure mode for an account-identity migration.
3. **Clean 4xx surfaced to the caller**: confirmed. `AuthController.register` catches `UserAlreadyExistsException` and maps it to `ConflictException` (409); nothing raw leaks to the client.
4. **CORE/Integration boundary + repository error-conversion pattern**: clean. `UserRepositoryPort`'s signature is untouched, normalization is a private method internal to the adapter, and the `QueryFailedError` -> domain-exception conversion follows the documented pattern in `docs/architecture-overview.md`/`docs/engineering-standards.md`.
5. **Test coverage - partial gap**: see IMPORTANT below.
6. **No regression on the customer-identity email-hash collision system**: confirmed clean. `grep` for `findByEmail` shows only two callers (`RegistrationService`, `PasswordResetService`), both correctly benefiting from the centralized normalization. `CustomerProjectionRepositoryPort.findByEmailHash` (the separate collision-policy system documented in the architecture overview) is untouched and unrelated.

### IMPORTANT

**No test actually exercises the case-insensitive collision through a real Postgres unique constraint.** The new `user.repository.spec.ts` is solid (normalization on `findByEmail`/`save`, null passthrough, and the `23505` conversion), but the `23505` case is a hand-constructed `QueryFailedError` against a fully mocked `ormRepository` - it verifies the *catch* logic, not that Postgres's actual `UQ_users_email` constraint on the now-lowercased column truly rejects a case-variant duplicate end to end. `registration.service.spec.ts` and `password-reset.service.spec.ts` are unmodified in this diff, even though both the PR's own test-plan checklist and the linked implementation plan (`docs/plans/implementation-plan-unique-email-enforcement.md`, Testing Strategy section) call for a case-variant regression case in each. Since `registration.service.spec.ts` mocks `UserRepositoryPort` directly, it can't exercise the real normalization path either way - the only place that could verify the true end-to-end scenario ("register `User@Test.com` after `user@test.com` already exists" against a real DB) is an integration test, and none exists for the registration flow (there's no `apps/api/test/integration/**/*regist*` or `*user*` int-spec at all today). Given this is a security-labeled issue about account-identity uniqueness, recommend adding one focused `*.int-spec.ts` assertion (or at minimum wiring the case-variant test into `registration.service.spec.ts` against the real `UserRepository` rather than a mocked port) before merge, so the guarantee is verified against real Postgres behavior rather than a simulated error object.

### SUGGESTION

- The PR description's own checklist items for the FE test and migration dry-run are left unchecked (`- [ ] pnpm --filter @openlinker/web test...`, `- [ ] pnpm --filter @openlinker/api migration:run against a fresh DB`). Worth confirming these were actually run before merge, since the FE test file is present in the diff (so likely just an unchecked box), but the fresh-DB migration run is the one case that would catch an ordering/collision issue the reviewer can't fully verify from the diff alone.
- Minor, not introduced by this PR: `UserAlreadyExistsException`'s message embeds the raw normalized email (`User already exists: ${identifier}`), which `ConflictException(error.message)` forwards to the client verbatim. This is pre-existing behavior unrelated to this change, just flagging since this PR is the one making email the identity key more strictly - not a blocker.

### Positive notes

- Correctly rejected `citext` and a functional index in favor of normalize-on-write, with the trade-off reasoning documented directly in the plan/analysis docs shipped in this PR - good, proportionate decision for a codebase where all writes already funnel through one repository.
- The migration's self-documenting header (idempotency, irreversibility rationale, failure-mode-on-collision) is a good example of the "document complex migrations with inline comments" practice from `docs/migrations.md`.
- Timestamp `1819000000000` correctly sorts after the current tip (`1818000000007`), satisfying the ordering invariant.
- `ApiError.isConflict()` mirrors the existing `isUnauthorized()/isForbidden()/isNotFound()` sibling pattern exactly - no new pattern introduced.

### Verdict

No BLOCKING issues. One IMPORTANT item (test coverage for the actual DB-enforced case-insensitive collision) should be addressed before merge given the security label on the underlying issue, but the mechanism itself (DB-level normalize-on-write + existing unique constraint + existing error-conversion pattern) is correctly implemented.

File: libs/core/src/users/infrastructure/persistence/repositories/user.repository.ts [hunk 1]
```diff
@@ -33,7 +33,9 @@ export class UserRepository implements UserRepositoryPort {
   }
 
   async findByEmail(email: string): Promise<User | null> {
-    const entity = await this.ormRepository.findOne({ where: { email } });
+    const entity = await this.ormRepository.findOne({
+      where: { email: email.trim().toLowerCase() },
```
Comment (piotrswierzy) [inline libs/core/src/users/infrastructure/persistence/repositories/user.repository.ts:37]: **SUGGESTION** — DRY: `findByEmail` inlines `email.trim().toLowerCase()`, while `save` delegates to the new private `normalizeEmail(...)` helper (line ~155). Two copies of the same normalization rule is exactly the kind of drift that reintroduces the bug this PR closes (e.g. someone later adds Unicode folding to one path but not the other). Route the read path through the same helper: `this.normalizeEmail(email) ?? ''` (or widen `normalizeEmail` for the non-null case). Single source of truth for the canonical form. Not blocking — the two implementations are byte-identical today.

File: apps/web/src/features/users/hooks/use-register-mutation.ts [hunk 2]
```diff
@@ -1,9 +1,10 @@
 import { useMutation, type UseMutationResult } from '@tanstack/react-query';
 import { useApiClient } from '../../../app/api/api-client-provider';
+import type { ApiError } from '../../../shared/api/api-error';
 import type { RegisterRequest } from '../../auth/api/auth.types';
 import type { OkResponse } from '../../auth/api/auth.types';
 
-export function useRegisterMutation(): UseMutationResult<OkResponse, Error, RegisterRequest> {
+export function useRegisterMutation(): UseMutationResult<OkResponse, ApiError, RegisterRequest> {
```
Comment (piotrswierzy) [inline apps/web/src/features/users/hooks/use-register-mutation.ts:7]: **SUGGESTION** — The error channel is now typed `ApiError`, but the mutation can still reject with a plain `Error` (e.g. a network/parse failure that never becomes an `ApiError`). The component correctly defends with `register.error instanceof ApiError && register.error.isConflict()` in `register-form.tsx`, so this is safe at the call site — but the type now over-promises what the mutation guarantees. Consider keeping `Error` as the declared error type (the `instanceof ApiError` narrowing already gives you `isConflict()` where needed), or confirm the api-client contract truly only ever rejects with `ApiError`. Cosmetic; no behavior impact.

**Review by piotrswierzy — APPROVED** [review]
## Review — `feat(users): enforce case-insensitive unique email on registration`

**Verdict: APPROVE.** Correct, well-scoped, and appropriately conservative. No blockers.

> ⚠️ **Stacked PR** — base branch is `1606-demo-mode-improvements`, not `main`. Reviewed against its own base (`get_diff`). This is intentional; merge order should respect the stack (parent #1606 first). Not a defect.

Part of #1606, closes #1625.

### What it does
Closes the one real gap in an already-implemented uniqueness path (DB `UQ_users_email` constraint + service pre-check + repository `23505`→domain-exception backstop all pre-existed): **case sensitivity**. It normalizes `users.email` to trimmed-lowercase on both write (`UserRepository.save`) and read (`findByEmail`), adds a data-backfill migration for pre-existing mixed-case rows, and surfaces a dedicated "This email is already registered." message on the register form for 409s.

### Assessment
- **Mechanism (correct).** Normalize-on-write + the existing case-sensitive `UQ_users_email` unique constraint = effective case-insensitive uniqueness (all stored values become lowercase, so the plain constraint enforces it). Reads normalize with the identical rule, so lookups match writes — verified `RegistrationService.register` pre-check (`registration.service.ts:62`) and `PasswordResetService.requestReset` (`password-reset.service.ts:50`) both go through `findByEmail` and therefore inherit the normalization. Chosen over `citext` / functional index with a documented, defensible rationale (no new extension, no schema-type churn).
- **Migration (present & correct).** `1819000000000-normalize-user-emails.ts` — a data backfill (no ORM schema change, so no DDL migration was owed; the data migration for existing rows is the right call and it's here). Timestamp `1819000000000` is strictly greater than main's tail `1818000000007` ✓ (`docs/migrations.md` ordering). Idempotent `WHERE "email" <> LOWER(TRIM("email"))`. `down()` intentionally irreversible, documented.
- **Existing-row collision risk (handled).** Two rows that normalize to the same value make the `UPDATE` trip Postgres `23505` and fail the migration loudly rather than silently merge/drop an account — the correct, data-loss-avoidant behavior, documented in the migration header for the operator.
- **Validation / layering / error handling.** `RegisterDto` already carries `@IsEmail()` (public endpoint input validation intact); `UserAlreadyExistsException`→`ConflictException` 409 mapping (`auth.controller.ts:130`) unchanged; no `any`; explicit return types on `normalizeEmail`/`isConflict`; no CORE↔Integration boundary touched; `UserRepositoryPort` signature unchanged. `+` sub-addressing is correctly **not** stripped (it's delivery-significant; distinct from the customers-context Allegro-mask stripping).
- **Tests.** Good coverage: normalized read, whitespace trim, lowercased persist, null-email passthrough, the `23505` race→`UserAlreadyExistsException` path, and non-unique `QueryFailedError` re-throw; FE `isConflict()` + 409-message unit tests. Whitespace/null/idempotency edge cases covered.

### BLOCKING
None.

### IMPORTANT
None.

### SUGGESTION
1. **`user.repository.ts:37`** — `findByEmail` inlines `email.trim().toLowerCase()` while `save` uses the new `normalizeEmail(...)` helper. Two copies of the canonical-form rule invites future drift (the exact class of bug this PR fixes). Route both through the one helper. (inline comment)
2. **`use-register-mutation.ts:7`** — error channel retyped `ApiError`, but the mutation may still reject with a plain `Error`; the component's `instanceof ApiError` guard keeps it safe, so this is cosmetic over-promising only. (inline comment)
3. **Deploy ordering (non-blocking note).** Between shipping the normalizing code and running the backfill migration, an existing mixed-case user's login/reset lookup won't resolve (query normalizes, stored row hasn't yet). Standard "migrate + deploy together" resolves it; the plan already calls this out — flagging for the operator only.

### Positives
Tight scope with an honest "most of this already existed" framing; reads and writes share one normalization rule; fail-loud (not merge-loud) migration on genuine collisions; thorough unit tests including the race path; correct decision to leave `+` addressing and Unicode/IDN out of scope. Nicely reasoned plan + pre-implement analysis included.

**Comment by norbert-kulus-blockydevs** [comment]
Addressed the test coverage gap flagged in the review.

**New: a real-Postgres integration test.** Added `apps/api/test/integration/user-email-uniqueness.int-spec.ts`. It resolves the real `UserRepositoryPort` from the app's DI container (via `USER_REPOSITORY_TOKEN`) and calls `save()` directly twice - first with `user@test.com`, then with `User@Test.com` - bypassing `RegistrationService.findByEmail`'s pre-check entirely, so the second `save()` call is the only thing standing between a case-variant email and a duplicate row. It asserts the real Postgres `23505` unique-violation on `UQ_users_email` is converted to a genuine `UserAlreadyExistsException`, and confirms no duplicate row was persisted. A second case confirms two genuinely distinct emails still succeed. This runs against a real Testcontainers Postgres, not a mocked `QueryFailedError`.

**Filled the plan-mandated unit test gaps.** The implementation plan (`docs/plans/implementation-plan-unique-email-enforcement.md`) called for case-variant regression tests in `registration.service.spec.ts` and `password-reset.service.spec.ts`, but neither file had them:
- `registration.service.spec.ts`: added a test asserting `register('user2', 'FOO@EXAMPLE.COM', ...)` throws `UserAlreadyExistsException` when the repo already holds a case-variant match.
- `password-reset.service.spec.ts`: added a test asserting `requestReset('Admin@Example.COM')` resolves the existing user and triggers the notifier, mirroring how the real repository normalizes lookups.

**Quality gate (scoped to the touched areas):**
- `pnpm --filter @openlinker/api lint` - clean (pre-existing unrelated warning/error in `plugins.ts` and `password-reset.service.spec.ts` confirmed present before this change too, via stash-and-diff).
- `pnpm --filter @openlinker/api type-check` - clean.
- Unit tests scoped to `auth` + `users` (`apps/api`) - 189/189 passing.
- Unit tests scoped to `users` (`libs/core`) - 22/22 passing.
- New int-spec run in isolation against real Testcontainers Postgres - 2/2 passing.

Pushed to the branch; ready for another look.

**Comment by norbert-kulus-blockydevs** [comment]
Re-review after the fix for the prior IMPORTANT finding (real-Postgres coverage for the case-insensitive unique-email constraint). Verified by reading the actual diff (`gh pr diff 1641`), not the author's summary.

## Verification of the fix

**1. Real Postgres, real DI, real save() path - confirmed.**
`apps/api/test/integration/user-email-uniqueness.int-spec.ts` uses `getTestHarness()` from `apps/api/test/integration/setup.ts`, which wraps `@openlinker/test-kit`'s `createIntegrationTestHarness` - genuine Testcontainers-backed Postgres (`libs/test-kit/src/containers.ts` uses `PostgreSqlContainer` from `@testcontainers/postgresql`), matching this repo's `*.int-spec.ts` convention (`docs/testing-guide.md`).

The test resolves `USER_REPOSITORY_TOKEN` from `harness.getApp()` - a real `INestApplication`. Checked the binding in `libs/core/src/users/users.module.ts`: `{ provide: USER_REPOSITORY_TOKEN, useExisting: UserRepository }`. So this is the production `UserRepository.save()` implementation, hitting a real DB, not a hand-rolled repository or another mock. This closes the original gap exactly as intended.

**2. Tests the right thing - confirmed.**
Two `save()` calls: `user@test.com` then `User@Test.com`. The second is asserted to reject with `UserAlreadyExistsException` (the real domain exception the repository's `23505`-catch throws, not a raw `QueryFailedError` leak). The test also queries the DB directly afterward (`SELECT count(*) ... WHERE email = 'user@test.com'`) and asserts exactly one row - good extra rigor confirming no duplicate silently persisted.

Traced the full chain: `UserRepository.save` normalizes via the new private `normalizeEmail()` (trim+lowercase) before `ormRepository.create`, catches the TypeORM `QueryFailedError` with `code === '23505'`, and throws `UserAlreadyExistsException`. The `UQ_users_email` unique constraint (`apps/api/src/migrations/1775000000000-add-users-table.ts`) is on the raw column, so this only works because storage is normalized - correctly exercised end-to-end.

**3. Positive case (genuinely distinct emails) - confirmed, not tautological.**
Second test in the new int-spec: `carol@test.com` and `dave@test.com` both save successfully. This is a meaningful contrast case - it proves the constraint/exception path isn't firing spuriously for unrelated emails.

**4. New unit tests in `registration.service.spec.ts` / `password-reset.service.spec.ts` - useful, correctly scoped, not just duplicating the int-spec.**
These operate one layer up and add real value:
- `registration.service.spec.ts`: asserts `RegistrationService.register('user2', 'FOO@EXAMPLE.COM', ...)` throws `UserAlreadyExistsException` given a mocked `findByEmail` returning an existing user, and confirms `repo.save` is never called - this is the pre-check branch, distinct from the int-spec's DB-constraint-race branch. Together the two levels cover both halves of the documented defense-in-depth (pre-check + DB backstop).
- `password-reset.service.spec.ts`: asserts a mixed-case email lookup (`Admin@Example.COM`) still resolves the user and triggers `notifyResetRequested`. This is the third call site that depends on `findByEmail` normalizing internally - good regression coverage for a service that isn't touched by code changes but depends on the shared normalization.

Also independently confirmed the new `libs/core/src/users/infrastructure/persistence/repositories/user.repository.spec.ts` (not explicitly asked about, but part of the diff) unit-tests `normalizeEmail` behavior for `findByEmail`/`save` (mixed case, whitespace, null-passthrough) plus the `23505` conversion and a non-unique-violation passthrough case, using a properly mocked `Repository<UserOrmEntity>` via Nest's `Test.createTestingModule` + `getRepositoryToken` - correct unit-test pattern (mocks the ORM repository, not the port).

**5. No regressions in the previously-clean findings.**
- Atomicity/race-safety: normalize-before-insert (`normalizeEmail(user.email)` computed once, used both for the entity and for the `detail.includes('(email)')` check) + the DB unique constraint as the real backstop - unchanged design, now proven live.
- Migration: `1819000000000-normalize-user-emails.ts` sorts strictly after the confirmed tip `1818000000007-add-inventory-item-is-stale.ts` - ordering invariant holds. Fails loudly (23505) on genuine pre-existing case collisions - documented and intentional, consistent with the plan.
- Clean 4xx surfaced: `ApiError.isConflict()` (409) plus the register-form's dedicated "This email is already registered." message - unit- and component-tested.
- No CORE/Integration boundary issue: change is entirely inside `libs/core/src/users` infrastructure + its `apps/api/src/auth` host callers. `UserRepositoryPort` signature is unchanged.
- No accidental touch to the customer-identity email-hash collision system: confirmed no diff hunk under `libs/core/src/customers/**` - the two systems remain fully independent.

## Verdict

**Ready to merge.** The prior IMPORTANT finding is resolved: the new int-spec genuinely exercises a real Postgres unique-violation through the real DI-resolved repository and real `save()` path, with both a correctly-failing case-variant duplicate and a correctly-passing distinct-email case. The accompanying unit-test additions are appropriately scoped to their own service boundaries and add real incremental coverage rather than redundant duplication. No BLOCKING or IMPORTANT items remain. No SUGGESTIONs beyond what's already noted in the plan's own documented, intentional migration failure-mode.
